presto-hive/src/main/java/com/facebook/presto/hive/S3SelectRecordCursor.java

```diff
@@ -0,0 +1,201 @@
+        }
+        return Joiner.on(",").join(columns.stream().map(HiveColumnHandle::getName).collect(Collectors.toList()));
+    }
+
+    static String buildColumnTypes(List<HiveColumnHandle> columns)
+    {
+        if (columns == null || columns.isEmpty()) {
+            return EMPTY_STRING;
+        }
+        return Joiner.on(",").join(columns.stream()
+                .map(column -> column.getHiveType().getTypeInfo().getTypeName()).collect(Collectors.toList()));
+    }
+
+    static ThriftDDL parseThriftDDL(String ddl)
+    {
+        if (Strings.isNullOrEmpty(ddl)) {
+            return null;
+        }
+        String[] parts = ddl.trim().split("\\s+");
+        Preconditions.checkArgument(parts.length >= 5, "Invalid thrift DDL " + ddl);
+        Preconditions.checkArgument(THRIFT_STRUCT.equals(parts[0]), "ThriftDDL should start with " + THRIFT_STRUCT);
+        ThriftDDL thriftDDL = new ThriftDDL();
+        thriftDDL.setTableName(parts[1]);
+        Preconditions.checkArgument(START_STRUCT.equals(parts[2]), "Invalid thrift DDL " + ddl);
+        Preconditions.checkArgument(parts[parts.length - 1].endsWith(END_STRUCT), "Invalid thrift DDL " + ddl);
+        parts[parts.length - 1] = parts[parts.length - 1].substring(0, parts[parts.length - 1].length() - 1);
+        List<ThriftField> fields = new ArrayList<>();
+        for (int i = 3; i < parts.length - 1; i += 2) {
+            ThriftField thriftField = new ThriftField();
+            thriftField.setType(parts[i]);
+            if (parts[i + 1].endsWith(FIELD_SEPARATOR)) {
+                parts[i + 1] = parts[i + 1].substring(0, parts[i + 1].length() - 1);
+            }
+            thriftField.setName(parts[i + 1]);
+            fields.add(thriftField);
+        }
+        thriftDDL.setFields(fields);
+
+        return thriftDDL;
+    }
+
+    static ThriftDDL filterThriftDDL(ThriftDDL thriftDDL, List<HiveColumnHandle> columns)
+    {
+        if (thriftDDL == null || thriftDDL.getFields() == null || thriftDDL.getFields().isEmpty()) {
+            return thriftDDL;
+        }
+        Set<String> columnNames = new HashSet<>();
+        for (HiveColumnHandle column : columns) {
+            columnNames.add(column.getName());
+        }
+        thriftDDL.setFields(thriftDDL.getFields().stream().filter(field -> columnNames.contains(field.getName())).collect(Collectors.toList()));
+
+        return thriftDDL;
+    }
+
+    static String convString(ThriftDDL thriftDDL)
+    {
+        if (thriftDDL == null || thriftDDL.getFields() == null || thriftDDL.getFields().isEmpty()) {
+            return EMPTY_STRING;
+        }
+        StringBuilder sb = new StringBuilder();
+        sb.append(THRIFT_STRUCT).append(SPACE).append(thriftDDL.getTableName()).append(SPACE).append(START_STRUCT);
+        sb.append(Joiner.on(",").join(thriftDDL.getFields().stream()
+                .map(field -> SPACE + field.getType() + SPACE + field.getName()).collect(Collectors.toList())));
```

Comment:
this list of stream operations is hard to read.

Reply:
made more readable